File: 2024/day_2/Part2.java
import static java.util.function.Predicate.not;
import java.io.File;
import java.io.FileNotFoundException;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;
import java.util.Scanner;
import java.util.stream.Collectors;

public class Part2 {
  private static Scanner readInput(String[] args) throws FileNotFoundException {
    if (args.length == 0) return new Scanner(System.in);
    return new Scanner(new File(args[0]));
  }

  private static boolean isSafe(List<Integer> levels) {
    boolean ascending = levels.get(0) < levels.get(1);
    for (int i = 1; i < levels.size(); i++) {
      if (ascending && levels.get(i) < levels.get(i - 1)) return false;
      else if (!ascending && levels.get(i) > levels.get(i - 1)) return false;
    }

    for (int i = 1; i < levels.size(); i++) {
      int diff = Math.abs(levels.get(i) - levels.get(i - 1));
      if (diff < 1 || diff > 3) return false;
    }
    return true;
  }

  public static void main(String[] args) throws FileNotFoundException {
    System.out.println(readInput(args).useDelimiter("\n").tokens().filter(not(String::isBlank)).filter(line -> {
      List<Integer> levels =
          Arrays.stream(line.split("\\s+")).map(Integer::parseInt).collect(Collectors.toCollection(LinkedList::new));
      if (isSafe(levels)) return true;
      for (int i = 0; i < levels.size(); i++) {
        int l = levels.remove(i);
        if (isSafe(levels)) return true;
        levels.add(i, l);
      }
      return false;
    }).count());
  }

}

File: 2024/day_2/part2.ts
// https://adventofcode.com/2024/day/1

import { readInput } from "../../utils/utils";

const input = await readInput();
const lines = input.split("\n").filter(Boolean);
const reports = lines.map(line => line.split(/\s+/).map(Number));

function isSafe(levels: number[]): boolean {
  if (levels[0] > levels[1]) {
    if (levels.some((level, i) => level < levels[i + 1])) return false;
  } else {
    if (levels.some((level, i) => level > levels[i + 1])) return false;
  }

  if (
    levels.some(
      (l, i) =>
        Math.abs(l - levels[i + 1]) < 1 || Math.abs(l - levels[i + 1]) > 3
    )
  )
    return false;
  return true;
}

console.log(
  reports.filter(levels => {
    if (isSafe(levels)) return true;
    for (let i = 0; i < levels.length; i++) {
      const [l] = levels.splice(i, 1);
      if (isSafe(levels)) return true;
      levels.splice(i, 0, l);
    }
    return false;
  }).length
);
